File: crates/viben-core/src/cli/commands/cron.rs
//! viben cron command
//!
//! CLI commands for managing scheduled tasks (cron jobs).

use clap::{Args, Subcommand};
use serde_json::json;
use std::sync::Arc;

use crate::cli::{
    CliContext,
    error::{CliError, CliResult},
    output::{print_json, print_simple_table, print_success, print_error, SuccessResponse},
};
use crate::services::{CronService, CreateCronJob, EventService};

#[derive(Args)]
pub struct CronCommand {
    #[command(subcommand)]
    pub action: CronAction,
}

#[derive(Subcommand)]
pub enum CronAction {
    /// List all cron jobs
    List,
    /// Add a new cron job
    Add {
        /// Job name
        name: String,
        /// Cron schedule (e.g., "0 * * * *") - use this OR --every, not both
        #[arg(short, long)]
        schedule: Option<String>,
        /// Interval in seconds (e.g., 60 for every minute) - use this OR --schedule
        #[arg(short, long)]
        every: Option<u64>,
        /// Message/command to execute
        #[arg(short, long)]
        message: String,
        /// Agent ID to use (default: main)
        #[arg(short, long)]
        agent: Option<String>,
        /// Custom job ID (auto-generated from name if not provided)
        #[arg(long)]
        id: Option<String>,
        /// Start disabled
        #[arg(long)]
        disabled: bool,
    },
    /// Remove a cron job
    Remove {
        /// Job ID
        id: String,
    },
    /// Enable a cron job
    Enable {
        /// Job ID
        id: String,
    },
    /// Disable a cron job
    Disable {
        /// Job ID
        id: String,
    },
    /// Run a cron job immediately
    Run {
        /// Job ID
        id: String,
    },
    /// Show cron job details
    Show {
        /// Job ID
        id: String,
    },
    /// Start the cron scheduler (keeps running until interrupted)
    Start {
        /// Run in foreground with verbose output
        #[arg(short, long)]
        verbose: bool,
    },
}

impl CronCommand {
    pub async fn execute(self, ctx: CliContext) -> CliResult<()> {
        // Create event service and cron service
        let events = Arc::new(EventService::new());
        let cron_service = CronService::new(events);

        match self.action {
            CronAction::List => {
                // Load existing jobs from config
                let _ = cron_service.start().await;
                let jobs = cron_service.list_jobs().await;
                cron_service.shutdown().await;

                if ctx.json {
                    print_json(&SuccessResponse::new(json!({ "jobs": jobs })));
                } else if jobs.is_empty() {
                    println!("No cron jobs configured");
                    println!("\nUse 'viben cron add' to create a new job");
                } else {
                    let headers = &["ID", "NAME", "SCHEDULE", "ENABLED", "LAST RUN", "NEXT RUN"];
                    let rows: Vec<Vec<String>> = jobs
                        .iter()
                        .map(|j| {
                            let schedule = if let Some(ref cron) = j.cron {
                                cron.clone()
                            } else if let Some(every) = j.every {
                                format!("every {}s", every)
                            } else {
                                "none".to_string()
                            };
                            let enabled = if j.enabled { "yes" } else { "no" };
                            let last_run = j
                                .last_run
                                .map(|ts| {
                                    chrono::DateTime::from_timestamp_millis(ts)
                                        .map(|dt| dt.format("%Y-%m-%d %H:%M:%S").to_string())
                                        .unwrap_or_else(|| "-".to_string())
                                })
                                .unwrap_or_else(|| "-".to_string());
                            let next_run = j
                                .next_run
                                .map(|ts| {
                                    chrono::DateTime::from_timestamp_millis(ts)
                                        .map(|dt| dt.format("%Y-%m-%d %H:%M:%S").to_string())
                                        .unwrap_or_else(|| "-".to_string())
                                })
                                .unwrap_or_else(|| "-".to_string());
                            vec![
                                j.id.clone(),
                                j.name.clone(),
                                schedule,
                                enabled.to_string(),
                                last_run,
                                next_run,
                            ]
                        })
                        .collect();
                    print_simple_table(headers, &rows);
                }
            }

            CronAction::Add {
                name,
                schedule,
                every,
                message,
                agent,
                id,
                disabled,
            } => {
                // Validate: must have either schedule or every, but not both
                if schedule.is_none() && every.is_none() {
                    print_error("Must specify either --schedule (cron expression) or --every (interval in seconds)");
                    return Err(CliError::InvalidArgument(
                        "Must specify either --schedule or --every".to_string(),
                    ));
                }
                if schedule.is_some() && every.is_some() {
                    print_error("Cannot specify both --schedule and --every");
                    return Err(CliError::InvalidArgument(
                        "Cannot specify both --schedule and --every".to_string(),
                    ));
                }

                let create = CreateCronJob {
                    id,
                    name: name.clone(),
                    message,
                    cron: schedule,
                    every,
                    channel: None,
                    agent,
                    enabled: !disabled,
                };

                match cron_service.create_job(create).await {
                    Ok(job) => {
                        if ctx.json {
                            print_json(&SuccessResponse::new(json!(job)));
                        } else {
                            print_success(&format!("Created cron job: {} (ID: {})", job.name, job.id));
                            if let Some(next) = job.next_run {
                                if let Some(dt) = chrono::DateTime::from_timestamp_millis(next) {
                                    println!("  Next run: {}", dt.format("%Y-%m-%d %H:%M:%S UTC"));
                                }
                            }
                            if !job.enabled {
                                println!("  Status: disabled (use 'viben cron enable {}' to activate)", job.id);
                            }
                        }
                    }
                    Err(e) => {
                        if ctx.json {
                            print_json(&json!({ "error": e.to_string() }));
                        } else {
                            print_error(&format!("Failed to create job: {}", e));
                        }
                        return Err(CliError::Core(crate::Error::Agent(e.to_string())));
                    }
                }
            }

            CronAction::Remove { id } => {
                // Load jobs first
                let _ = cron_service.start().await;

                match cron_service.delete_job(&id).await {
                    Ok(()) => {
                        cron_service.shutdown().await;
                        if ctx.json {
                            print_json(&SuccessResponse::new(json!({ "removed": id })));
                        } else {
                            print_success(&format!("Removed cron job: {}", id));
                        }
                    }
                    Err(e) => {
                        cron_service.shutdown().await;
                        if ctx.json {
                            print_json(&json!({ "error": e.to_string() }));
                        } else {
                            print_error(&format!("Failed to remove job: {}", e));
                        }
                        return Err(CliError::Core(crate::Error::Agent(e.to_string())));
                    }
                }
            }

            CronAction::Enable { id } => {
                // Load jobs first
                let _ = cron_service.start().await;

                match cron_service.enable_job(&id).await {
                    Ok(job) => {
                        cron_service.shutdown().await;
                        if ctx.json {
                            print_json(&SuccessResponse::new(json!(job)));
                        } else {
                            print_success(&format!("Enabled cron job: {}", id));
                            if let Some(next) = job.next_run {
                                if let Some(dt) = chrono::DateTime::from_timestamp_millis(next) {
                                    println!("  Next run: {}", dt.format("%Y-%m-%d %H:%M:%S UTC"));
                                }
                            }
                        }
                    }
                    Err(e) => {
                        cron_service.shutdown().await;
                        if ctx.json {
                            print_json(&json!({ "error": e.to_string() }));
                        } else {
                            print_error(&format!("Failed to enable job: {}", e));
                        }
                        return Err(CliError::Core(crate::Error::Agent(e.to_string())));
                    }
                }
            }

            CronAction::Disable { id } => {
                // Load jobs first
                let _ = cron_service.start().await;

                match cron_service.disable_job(&id).await {
                    Ok(job) => {
                        cron_service.shutdown().await;
                        if ctx.json {
                            print_json(&SuccessResponse::new(json!(job)));
                        } else {
                            print_success(&format!("Disabled cron job: {}", id));
                        }
                    }
                    Err(e) => {
                        cron_service.shutdown().await;
                        if ctx.json {
                            print_json(&json!({ "error": e.to_string() }));
                        } else {
                            print_error(&format!("Failed to disable job: {}", e));
                        }
                        return Err(CliError::Core(crate::Error::Agent(e.to_string())));
                    }
                }
            }

            CronAction::Run { id } => {
                // Load jobs first
                let _ = cron_service.start().await;

                println!("Running job '{}' immediately...", id);

                match cron_service.run_job(&id).await {
                    Ok(()) => {
                        // Get updated job status
                        if let Some(job) = cron_service.get_job(&id).await {
                            cron_service.shutdown().await;
                            if ctx.json {
                                print_json(&SuccessResponse::new(json!(job)));
                            } else {
                                print_success(&format!("Job '{}' executed", id));
                                if let Some(status) = &job.last_status {
                                    println!("  Status: {}", status);
                                }
                                if let Some(error) = &job.last_error {
                                    println!("  Error: {}", error);
                                }
                            }
                        } else {
                            cron_service.shutdown().await;
                            print_success(&format!("Job '{}' executed", id));
                        }
                    }
                    Err(e) => {
                        cron_service.shutdown().await;
                        if ctx.json {
                            print_json(&json!({ "error": e.to_string() }));
                        } else {
                            print_error(&format!("Failed to run job: {}", e));
                        }
                        return Err(CliError::Core(crate::Error::Agent(e.to_string())));
                    }
                }
            }

            CronAction::Show { id } => {
                // Load jobs first
                let _ = cron_service.start().await;

                match cron_service.get_job(&id).await {
                    Some(job) => {
                        cron_service.shutdown().await;
                        if ctx.json {
                            print_json(&SuccessResponse::new(json!(job)));
                        } else {
                            println!("Cron Job: {}", job.id);
                            println!("  Name: {}", job.name);
                            println!("  Enabled: {}", if job.enabled { "yes" } else { "no" });
                            println!("  Message: {}", job.message);
                            println!("  Agent: {}", job.agent);

                            if let Some(ref cron) = job.cron {
                                println!("  Schedule: {} (cron)", cron);
                            } else if let Some(every) = job.every {
                                println!("  Schedule: every {} seconds", every);
                            }

                            if let Some(ref channel) = job.channel {
                                println!("  Channel: {}", channel);
                            }

                            if let Some(ts) = job.last_run {
                                if let Some(dt) = chrono::DateTime::from_timestamp_millis(ts) {
                                    println!("  Last Run: {}", dt.format("%Y-%m-%d %H:%M:%S UTC"));
                                }
                            } else {
                                println!("  Last Run: never");
                            }

                            if let Some(ref status) = job.last_status {
                                println!("  Last Status: {}", status);
                            }

                            if let Some(ref error) = job.last_error {
                                println!("  Last Error: {}", error);
                            }

                            if let Some(ts) = job.next_run {
                                if let Some(dt) = chrono::DateTime::from_timestamp_millis(ts) {
                                    println!("  Next Run: {}", dt.format("%Y-%m-%d %H:%M:%S UTC"));
                                }
                            }

                            if let Some(dt) = chrono::DateTime::from_timestamp_millis(job.created_at) {
                                println!("  Created: {}", dt.format("%Y-%m-%d %H:%M:%S UTC"));
                            }
                            if let Some(dt) = chrono::DateTime::from_timestamp_millis(job.updated_at) {
                                println!("  Updated: {}", dt.format("%Y-%m-%d %H:%M:%S UTC"));
                            }
                        }
                    }
                    None => {
                        cron_service.shutdown().await;
                        if ctx.json {
                            print_json(&json!({ "error": format!("Job not found: {}", id) }));
                        } else {
                            print_error(&format!("Job not found: {}", id));
                        }
                        return Err(CliError::Core(crate::Error::Agent(format!(
                            "Job not found: {}",
                            id
                        ))));
                    }
                }
            }

            CronAction::Start { verbose } => {
                println!("Starting cron scheduler...");
                println!("Press Ctrl+C to stop\n");

                // Start the cron service
                if let Err(e) = cron_service.start().await {
                    print_error(&format!("Failed to start cron service: {}", e));
                    return Err(CliError::Core(crate::Error::Agent(e.to_string())));
                }

                let jobs = cron_service.list_jobs().await;
                let enabled_count = jobs.iter().filter(|j| j.enabled).count();
                println!("Loaded {} jobs ({} enabled)", jobs.len(), enabled_count);

                if verbose {
                    for job in &jobs {
                        let status = if job.enabled { "enabled" } else { "disabled" };
                        let schedule = if let Some(ref cron) = job.cron {
                            format!("cron: {}", cron)
                        } else if let Some(every) = job.every {
                            format!("every: {}s", every)
                        } else {
                            "no schedule".to_string()
                        };
                        println!("  - {} ({}) [{}]", job.name, schedule, status);
                    }
                }

                println!("\nScheduler running. Waiting for jobs to trigger...");

                // Wait for Ctrl+C
                tokio::signal::ctrl_c().await.ok();

                println!("\nShutting down...");
                cron_service.shutdown().await;
                println!("Cron scheduler stopped.");
            }
        }
        Ok(())
    }
}

File: packages/core/src/cli/commands/cron.ts
/**
 * viben cron - Cron job management commands
 *
 * Provides commands for managing scheduled cron jobs.
 * Jobs can be agent-based (send message to agent) or script-based (execute bash).
 *
 * CLI uses ops functions directly for data operations.
 * Jobs are not scheduled until Gateway is running.
 */
import chalk from "chalk";
import type { Command } from "commander";
import type { OutputContext } from "../types";
import {
  output,
  successResponse,
  outputTable,
  outputKeyValue,
  handleCommandError,
} from "../lib";

// Import ops functions from cron module
import * as cronOps from "../../cron/ops";
import type { CronJob, CreateCronJob } from "../../cron/ops";

/**
 * Format job status for display
 */
function formatStatus(status?: string): string {
  switch (status) {
    case "success":
      return chalk.green("success");
    case "failure":
      return chalk.red("failure");
    case "running":
      return chalk.yellow("running");
    default:
      return chalk.gray("-");
  }
}

/**
 * Format enabled status for display
 */
function formatEnabled(enabled: boolean): string {
  return enabled ? chalk.green("enabled") : chalk.gray("disabled");
}

/**
 * Format timestamp for display
 */
function formatTimestamp(ts?: number): string {
  if (!ts) return chalk.gray("-");
  return new Date(ts).toLocaleString();
}

/**
 * Get output context from program options
 */
function getContext(program: Command): OutputContext {
  const opts = program.opts();
  return {
    json: opts.json || false,
    verbose: opts.verbose || false,
    quiet: opts.quiet || false,
  };
}

/**
 * List all cron jobs
 */
async function listJobs(ctx: OutputContext): Promise<void> {
  const result = await cronOps.listJobs();

  if (!result.success) {
    output(
      ctx,
      { success: false, error: { code: "LOAD_ERROR", message: result.error || "Unknown error" } },
      () => {
        console.error(chalk.red(`Failed to load jobs: ${result.error || "Unknown error"}`));
      }
    );
    return;
  }

  const jobs = result.jobs;

  output(
    ctx,
    successResponse({
      jobs: jobs.map((j) => ({
        id: j.id,
        name: j.name,
        enabled: j.enabled,
        job_type: j.job_type,
        schedule: j.cron || (j.every ? `every ${j.every}s` : null),
        agent: j.agent,
        last_status: j.last_status,
        next_run: j.next_run,
      })),
      count: jobs.length,
    }),
    () => {
      console.log(chalk.bold("Cron Jobs:"));
      console.log();

      if (jobs.length === 0) {
        console.log(chalk.gray("  No cron jobs configured."));
        console.log();
        console.log("Add a job with:");
        console.log(chalk.cyan("  viben cron add --name <name> --cron '0 9 * * *' --agent-id <agent>"));
        return;
      }

      const headers = ["ID", "Name", "Status", "Type", "Schedule", "Last Run", "Next Run"];
      const rows = jobs.map((j) => [
        j.id,
        j.name,
        formatEnabled(j.enabled),
        j.job_type,
        cronOps.formatSchedule(j),
        formatStatus(j.last_status),
        formatTimestamp(j.next_run),
      ]);

      outputTable(ctx, headers, rows);
    }
  );
}

/**
 * Show job details
 */
async function showJob(ctx: OutputContext, id: string): Promise<void> {
  const configPath = cronOps.getDefaultConfigPath();
  const result = await cronOps.getJob(configPath, id);

  if (!result.success || !result.job) {
    output(
      ctx,
      { success: false, error: { code: "NOT_FOUND", message: result.error || `Job not found: ${id}` } },
      () => {
        console.error(chalk.red(result.error || `Job not found: ${id}`));
      }
    );
    return;
  }

  const job = result.job;

  output(ctx, successResponse(job), () => {
    console.log(chalk.bold(`Job: ${job.name}`));
    console.log();

    const details: Record<string, string | number | boolean | undefined> = {
      ID: job.id,
      Name: job.name,
      Enabled: job.enabled ? "Yes" : "No",
      Type: job.job_type,
      Agent: job.agent,
    };

    if (job.cron) {
      details["Cron"] = job.cron;
    }
    if (job.every) {
      details["Interval"] = `${job.every} seconds`;
    }
    if (job.message) {
      details["Message"] = job.message;
    }
    if (job.script) {
      details["Script"] = job.script.length > 50 ? job.script.slice(0, 50) + "..." : job.script;
    }
    if (job.channel) {
      details["Channel"] = job.channel;
    }

    details["Created"] = formatTimestamp(job.created_at);
    details["Updated"] = formatTimestamp(job.updated_at);
    details["Last Run"] = formatTimestamp(job.last_run);
    details["Last Status"] = job.last_status || "-";
    details["Next Run"] = formatTimestamp(job.next_run);

    if (job.last_error) {
      details["Last Error"] = job.last_error;
    }
    if (job.last_output) {
      details["Last Output"] = job.last_output.length > 100
        ? job.last_output.slice(0, 100) + "..."
        : job.last_output;
    }

    outputKeyValue(ctx, details);
  });
}

/**
 * Add a new cron job
 */
async function addJob(
  ctx: OutputContext,
  name: string,
  options: {
    cron?: string;
    every?: string;
    agentId?: string;
    message?: string;
    script?: string;
    channel?: string;
  }
): Promise<void> {
  // Validate options
  if (!options.cron && !options.every) {
    output(
      ctx,
      { success: false, error: { code: "VALIDATION_ERROR", message: "Must specify either --cron or --every" } },
      () => {
        console.error(chalk.red("Error: Must specify either --cron or --every"));
        console.log();
        console.log("Examples:");
        console.log(chalk.cyan("  viben cron add --name 'Daily report' --cron '0 9 * * *' --agent-id main"));
        console.log(chalk.cyan("  viben cron add --name 'Hourly check' --every 3600 --agent-id monitor"));
      }
    );
    return;
  }

  const createJob: CreateCronJob = {
    name,
    agent: options.agentId || "main",
    job_type: options.script ? "script" : "agent",
    cron: options.cron,
    every: options.every ? parseInt(options.every, 10) : undefined,
    message: options.message,
    script: options.script,
    channel: options.channel,
    enabled: true,
  };

  const configPath = cronOps.getDefaultConfigPath();
  const result = await cronOps.createJob(configPath, createJob);

  if (!result.success || !result.job) {
    output(
      ctx,
      { success: false, error: { code: "CREATE_ERROR", message: result.error || "Unknown error" } },
      () => {
        console.error(chalk.red(`Failed to create job: ${result.error || "Unknown error"}`));
      }
    );
    return;
  }

  const job = result.job;

  output(ctx, successResponse(job), () => {
    console.log(chalk.green(`Created cron job: ${job.name}`));
    console.log(`  ID: ${job.id}`);
    console.log(`  Type: ${job.job_type}`);
    console.log(`  Schedule: ${cronOps.formatSchedule(job)}`);
    console.log(`  Agent: ${job.agent}`);
    console.log();
    console.log(chalk.yellow("Note: Jobs are scheduled when Gateway is running."));
  });
}

/**
 * Remove a cron job
 */
async function removeJob(ctx: OutputContext, id: string): Promise<void> {
  const configPath = cronOps.getDefaultConfigPath();

  // Check if exists
  const getResult = await cronOps.getJob(configPath, id);
  if (!getResult.success || !getResult.job) {
    output(
      ctx,
      { success: false, error: { code: "NOT_FOUND", message: `Job not found: ${id}` } },
      () => {
        console.error(chalk.red(`Job not found: ${id}`));
      }
    );
    return;
  }

  const job = getResult.job;
  const result = await cronOps.deleteJob(configPath, id);

  if (!result.success) {
    output(
      ctx,
      { success: false, error: { code: "DELETE_ERROR", message: result.error || "Unknown error" } },
      () => {
        console.error(chalk.red(`Failed to delete job: ${result.error || "Unknown error"}`));
      }
    );
    return;
  }

  output(
    ctx,
    successResponse({ id, deleted: true }),
    () => {
      console.log(chalk.green(`Removed cron job: ${job.name}`));
      console.log(`  ID: ${id}`);
    }
  );
}

/**
 * Enable a cron job
 */
async function enableJob(ctx: OutputContext, id: string): Promise<void> {
  const configPath = cronOps.getDefaultConfigPath();
  const result = await cronOps.enableJob(configPath, id);

  if (!result.success || !result.job) {
    output(
      ctx,
      { success: false, error: { code: "UPDATE_ERROR", message: result.error || `Failed to enable job: ${id}` } },
      () => {
        console.error(chalk.red(result.error || `Failed to enable job: ${id}`));
      }
    );
    return;
  }

  const job = result.job;

  output(ctx, successResponse(job), () => {
    console.log(chalk.green(`Enabled cron job: ${job.name}`));
    console.log(`  ID: ${id}`);
    console.log();
    console.log(chalk.yellow("Note: Scheduling takes effect when Gateway is running."));
  });
}

/**
 * Disable a cron job
 */
async function disableJob(ctx: OutputContext, id: string): Promise<void> {
  const configPath = cronOps.getDefaultConfigPath();
  const result = await cronOps.disableJob(configPath, id);

  if (!result.success || !result.job) {
    output(
      ctx,
      { success: false, error: { code: "UPDATE_ERROR", message: result.error || `Failed to disable job: ${id}` } },
      () => {
        console.error(chalk.red(result.error || `Failed to disable job: ${id}`));
      }
    );
    return;
  }

  const job = result.job;

  output(ctx, successResponse(job), () => {
    console.log(chalk.yellow(`Disabled cron job: ${job.name}`));
    console.log(`  ID: ${id}`);
  });
}

/**
 * Run a cron job immediately
 * Note: This requires the Gateway to be running
 */
async function runJob(ctx: OutputContext, id: string): Promise<void> {
  const configPath = cronOps.getDefaultConfigPath();

  // Check if exists
  const result = await cronOps.getJob(configPath, id);
  if (!result.success || !result.job) {
    output(
      ctx,
      { success: false, error: { code: "NOT_FOUND", message: `Job not found: ${id}` } },
      () => {
        console.error(chalk.red(`Job not found: ${id}`));
      }
    );
    return;
  }

  const job = result.job;

  // Try to trigger via Gateway API
  try {
    const response = await fetch(`http://127.0.0.1:18790/api/cron/${id}/run`, {
      method: "POST",
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({ error: "Unknown error" })) as { error?: string };
      const errorMsg = errorData.error || "Unknown error";
      output(
        ctx,
        { success: false, error: { code: "API_ERROR", message: errorMsg } },
        () => {
          console.error(chalk.red(`Failed to run job: ${errorMsg}`));
          console.log();
          console.log(chalk.yellow("Ensure the Gateway is running:"));
          console.log(chalk.cyan("  viben service start gateway"));
        }
      );
      return;
    }

    output(
      ctx,
      successResponse({ id, name: job.name, triggered: true }),
      () => {
        console.log(chalk.green(`Triggered job: ${job.name}`));
        console.log(`  ID: ${id}`);
        console.log();
        console.log("Check execution logs with:");
        console.log(chalk.cyan(`  viben cron logs ${id}`));
      }
    );
  } catch (error) {
    output(
      ctx,
      { success: false, error: { code: "CONNECTION_ERROR", message: "Failed to connect to Gateway" } },
      () => {
        console.error(chalk.red("Failed to connect to Gateway"));
        console.log();
        console.log(chalk.yellow("Ensure the Gateway is running:"));
        console.log(chalk.cyan("  viben service start gateway"));
      }
    );
  }
}

/**
 * Show execution logs for a cron job
 */
async function showLogs(
  ctx: OutputContext,
  id: string,
  options: { limit?: string; offset?: string }
): Promise<void> {
  const configPath = cronOps.getDefaultConfigPath();

  // Get job first
  const jobResult = await cronOps.getJob(configPath, id);
  if (!jobResult.success || !jobResult.job) {
    output(
      ctx,
      { success: false, error: { code: "NOT_FOUND", message: `Job not found: ${id}` } },
      () => {
        console.error(chalk.red(`Job not found: ${id}`));
      }
    );
    return;
  }

  const job = jobResult.job;
  const limit = options.limit ? parseInt(options.limit, 10) : 20;
  const offset = options.offset ? parseInt(options.offset, 10) : 0;

  const logsResult = await cronOps.getExecutionLogs(job, limit, offset);

  if (!logsResult.success) {
    output(
      ctx,
      { success: false, error: { code: "LOAD_ERROR", message: logsResult.error || "Unknown error" } },
      () => {
        console.error(chalk.red(`Failed to load logs: ${logsResult.error || "Unknown error"}`));
      }
    );
    return;
  }

  const logs = logsResult.logs;

  output(
    ctx,
    successResponse({ job_id: id, logs, count: logs.length }),
    () => {
      console.log(chalk.bold(`Execution Logs: ${job.name}`));
      console.log();

      if (logs.length === 0) {
        console.log(chalk.gray("  No execution logs found."));
        return;
      }

      const headers = ["Time", "Status", "Duration", "Trigger", "Output"];
      const rows = logs.map((log) => [
        new Date(log.started_at).toLocaleString(),
        log.status === "success"
          ? chalk.green(log.status)
          : log.status === "failure"
          ? chalk.red(log.status)
          : chalk.yellow(log.status),
        cronOps.formatDuration(log.duration_ms),
        log.trigger,
        log.error || (log.output ? log.output.slice(0, 50) + (log.output.length > 50 ? "..." : "") : "-"),
      ]);

      outputTable(ctx, headers, rows);
    }
  );
}

/**
 * Register the cron command
 */
export function registerCronCommand(program: Command): void {
  const cronCmd = program
    .command("cron")
    .description("Manage scheduled cron jobs");

  // cron list
  cronCmd
    .command("list")
    .description("List all cron jobs")
    .action(async () => {
      const ctx = getContext(program);
      try {
        await listJobs(ctx);
      } catch (error) {
        handleCommandError(ctx, error);
      }
    });

  // cron show <id>
  cronCmd
    .command("show")
    .argument("<id>", "Job ID")
    .description("Show job details")
    .action(async (id: string) => {
      const ctx = getContext(program);
      try {
        await showJob(ctx, id);
      } catch (error) {
        handleCommandError(ctx, error);
      }
    });

  // cron add --name <name>
  cronCmd
    .command("add")
    .requiredOption("-n, --name <name>", "Job name")
    .option("--cron <expression>", "Cron expression (e.g., '0 9 * * *')")
    .option("--every <seconds>", "Interval in seconds")
    .option("--agent-id <id>", "Agent ID to use", "main")
    .option("--message <message>", "Message to send to agent")
    .option("--script <bash>", "Bash script to execute (for script type)")
    .option("--channel <id>", "Channel ID to target")
    .description("Add a new cron job")
    .action(
      async (options: {
        name: string;
        cron?: string;
        every?: string;
        agentId?: string;
        message?: string;
        script?: string;
        channel?: string;
      }) => {
        const ctx = getContext(program);
        try {
          await addJob(ctx, options.name, options);
        } catch (error) {
          handleCommandError(ctx, error);
        }
      }
    );

  // cron remove <id>
  cronCmd
    .command("remove")
    .argument("<id>", "Job ID")
    .description("Remove a cron job")
    .action(async (id: string) => {
      const ctx = getContext(program);
      try {
        await removeJob(ctx, id);
      } catch (error) {
        handleCommandError(ctx, error);
      }
    });

  // cron enable <id>
  cronCmd
    .command("enable")
    .argument("<id>", "Job ID")
    .description("Enable a cron job")
    .action(async (id: string) => {
      const ctx = getContext(program);
      try {
        await enableJob(ctx, id);
      } catch (error) {
        handleCommandError(ctx, error);
      }
    });

  // cron disable <id>
  cronCmd
    .command("disable")
    .argument("<id>", "Job ID")
    .description("Disable a cron job")
    .action(async (id: string) => {
      const ctx = getContext(program);
      try {
        await disableJob(ctx, id);
      } catch (error) {
        handleCommandError(ctx, error);
      }
    });

  // cron run <id>
  cronCmd
    .command("run")
    .argument("<id>", "Job ID")
    .description("Run a job immediately (requires Gateway)")
    .action(async (id: string) => {
      const ctx = getContext(program);
      try {
        await runJob(ctx, id);
      } catch (error) {
        handleCommandError(ctx, error);
      }
    });

  // cron logs <id>
  cronCmd
    .command("logs")
    .argument("<id>", "Job ID")
    .option("-l, --limit <count>", "Number of logs to show", "20")
    .option("-o, --offset <count>", "Offset from most recent", "0")
    .description("Show execution logs for a job")
    .action(async (id: string, options: { limit?: string; offset?: string }) => {
      const ctx = getContext(program);
      try {
        await showLogs(ctx, id, options);
      } catch (error) {
        handleCommandError(ctx, error);
      }
    });
}
